backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHContextApi.scala

```diff
@@ -0,0 +1,130 @@
+package io.glutenproject.backendsapi.clickhouse
+
+import io.glutenproject.GlutenConfig
+import io.glutenproject.backendsapi.ContextApi
+import io.glutenproject.execution.CHBroadcastBuildSideRDD
+import io.glutenproject.vectorized.{CHNativeExpressionEvaluator, JniLibLoader}
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+
+import org.apache.commons.lang3.StringUtils
+import org.sparkproject.guava.cache.{Cache, CacheBuilder}
+
+import java.util
+import java.util.TimeZone
+import java.util.concurrent.TimeUnit
+
+class CHContextApi extends ContextApi with Logging {
+
+  override def initialize(conf: SparkConf): Unit = {
+    val libPath = conf.get(GlutenConfig.GLUTEN_LIB_PATH, StringUtils.EMPTY)
+    if (StringUtils.isBlank(libPath)) {
+      throw new IllegalArgumentException(
+        "Please set spark.gluten.sql.columnar.libpath to enable clickhouse backend")
+    }
+    // Path based load. Ignore all other loadees.
+    JniLibLoader.loadFromPath(libPath, true)
+
+    // Add configs
+    conf.set(
+      s"${CHBackendSettings.getBackendConfigPrefix()}.runtime_config.timezone",
+      conf.get("spark.sql.session.timeZone", TimeZone.getDefault.getID))
+    val initKernel = new CHNativeExpressionEvaluator()
+    initKernel.initNative(conf)
+  }
+
+  override def shutdown(): Unit = {
+    val kernel = new CHNativeExpressionEvaluator()
+    kernel.finalizeNative()
+  }
+
+  /**
+   * Store execution id for resource relations when execution start.
+   *
+   * @param executionId
+   *   execution id
+   */
+  override def onExecutionStart(executionId: String): Unit = {
+    if (executionId == null) {
+      logWarning(s"Execution Id is null. Resources maybe not clean after execution end.")
+    }
+  }
+
+  /**
+   * Clean up resource when execution end.
+   *
+   * @param executionId
+   *   execution id
+   */
+  override def onExecutionEnd(executionId: String): Unit = {
+    if (executionId == null) {
+      logWarning(s"Execution id $executionId not found. Can't clean resource right now")
+    } else {
+      val resources = CHContextApi.executionResourceRelation.getIfPresent(executionId)
+      if (resources != null) {
+        // clean broadcast hash data
+        resources.forEach(
+          resource_id => {
+            val v = CHBroadcastBuildSideRDD.buildSideRelationCache.getIfPresent(resource_id)
+            if (v != null) {
+              // Cleanup operations on the backend are idempotent
+              CHBroadcastBuildSideRDD.cleanBuildHashTable(resource_id, v.toLong)
+              CHBroadcastBuildSideRDD.buildSideRelationCache.invalidate(resource_id)
+            }
+          })
+        CHContextApi.executionResourceRelation.invalidate(executionId)
+      }
+    }
+  }
+
+  /**
+   * Create relation bhj broadcast data to execution id.
+   * @param executionId
+   *   execution id
+   * @param buildHashTableId
+   *   broadcast hashtable id
+   */
+  override def collectExecutionBroadcastHashTableId(
```

Comment:
1. Is `collectExecutionBroadcastHashTableId` Thread  Safe? 
2. Does it run only on Driver?

If so, please update comments.

Reply:
Already add comments